### PR TITLE
Add $max update operator compatibility tests

### PR DESCRIPTION
**Pull request by imforster**
#394 
Feature being tested: $max Update Operator
Number of test cases: 91
Link to MongoDB 8.2 documentation: https://www.mongodb.com/docs/v8.2/reference/operator/update/max/

Tests cover $max operator-specific behavior across 8 files (91 tests):

- Core behavior, null handling, type preservation, numeric comparisons
- Field paths (dot notation, field creation, multiple fields)
- Update result metadata (n/nModified) and upsert behavior
- Array positional operator wiring ($, $[], $[elem])
- Error cases (operator conflicts, dollar-prefixed fields)
- Edge cases (indexed fields, large documents)
- BSON cross-type wiring (2 representative cases per §19)
**Also adds:**
- UpdateTestCase dataclass at update/utils/ for parametrized update tests
- CONFLICTING_UPDATE_OPERATORS_ERROR and DOLLAR_PREFIXED_FIELD_NAME_ERROR constants

### CODE REVIEW

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_bson_wiring.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+from documentdb_tests.compatibility.tests.core.operator.update.utils import UpdateTestCase
+from documentdb_tests.framework.assertions import assertSuccess
+from documentdb_tests.framework.executor import execute_command
+from documentdb_tests.framework.parametrize import pytest_params
+
+_TEST_DATE = datetime(2023, 6, 1, tzinfo=timezone.utc)
+_TIMESTAMP = Timestamp(1672531200, 1)
+
+TESTS: list[UpdateTestCase] = [
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_bson_wiring.py:18]: 🔴 Critical — test_max_bson_wiring.py is the full BSON type-pair matrix, not a wiring sample (TEST_COVERAGE.md
  §19 violation)

  §19 Foundational Spec Behaviors — Test Once explicitly names this:

  ▎ BSON type ordering → tests/core/bson_types/. Operators that use it (e.g. $max, $gt, $sort) get 1-2 wiring
  ▎ cases, not the full type-pair matrix.

  The current file (306 lines, 33 cases) walks through cross-type comparisons across the full BSON ordering
  ladder: Null→Number, Number→String, String→Object, Object→Array, String→Binary, Binary→ObjectId,
  ObjectId→Boolean, Boolean→Date, Date→Timestamp, Timestamp→Regex, MinKey/MaxKey, plus array/object element-wise
  ordering and nested-array recursion. Every one of those rules is BSON ordering, not a $max-specific behavior.
  The agent's name suffix _bson_wiring.py per §19 says "small, representative, explicitly named" — citing
  tests/core/operator/expressions/comparisons/gt/test_gt_bson_wiring.py as the right shape.

  [Required] Cut to 1-2 wiring cases:
  - One upward-transition case (e.g., null_to_number_updates).
  - One downward-transition case (e.g., string_vs_objectid_unchanged).

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        {"update": collection.name, "updates": [{"q": test.query, "u": test.update}]},
+    )
+
+    result = execute_command(
+        collection, {"find": collection.name, "filter": {"_id": test.expected["_id"]}}
+    )
+    assertSuccess(result, [test.expected], msg=test.msg)
+
+
+def test_max_collation_case_insensitive(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py]: 🔴 Critical — test_max_collation_case_insensitive duplicates collation/$max coverage (FOLDER_STRUCTURE.md Rule
  2 + §19)

  Per Rule 2: "Cross-cutting features (rbac, transactions, collation, namespace rules, etc.) → own top-level
  folders, NOT under feature-specific folders." And §19: "Collation comparison → tests/core/collation/. Commands
  that accept collation test syntactic acceptance only. Sub-fields testing and semantic behavior is in
  tests/core/collation/."

  tests/core/collation/command_level/operations/test_operations_update_field_ops.py already has comprehensive
  $max + collation coverage — the existing COLLATION_UPDATE_MAX_TESTS matrix covers the case-insensitive
  equality scenario this PR adds (line 258).

  [Required] Drop test_max_collation_case_insensitive (test_max_comparisons.py:258-278). It duplicates existing
  collation-folder coverage.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+_EPOCH = datetime(1970, 1, 1, tzinfo=timezone.utc)
+_EARLY = datetime(2023, 1, 1, tzinfo=timezone.utc)
+_MID = datetime(2023, 6, 15, tzinfo=timezone.utc)
+_LATE = datetime(2025, 12, 31, tzinfo=timezone.utc)
+_FAR_FUTURE = datetime(9999, 12, 31, tzinfo=timezone.utc)
+
+TESTS: list[UpdateTestCase] = [
+    # Core behavior
+    UpdateTestCase(
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py:24]: 🟠 Major — String comparison edge cases overlap with BSON ordering

  In test_max_comparisons.py "String comparisons" group (lines 132-196): string_greater_updates,
  string_less_unchanged, empty_string_to_nonempty_updates, string_lexicographic_greater,
  string_prefix_unchanged, uppercase_vs_lowercase, unicode_string_comparison, very_long_string_updates. These
  test string-vs-string comparison rules — that's BSON string ordering, again per §19. Same cut argument: keep
  one wiring case ("string greater updates")

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_update_methods.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
+"""
+Update method and upsert tests for $max update field operator.
+
+Tests $max via updateOne, updateMany, findAndModify, bulkWrite,
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_update_methods.py]: Not true, for operators we only use one container feature.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_array_positional.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+"""
+
+import pytest
+
+from documentdb_tests.compatibility.tests.core.operator.update.utils import UpdateTestCase
+from documentdb_tests.framework.assertions import assertSuccess, assertSuccessPartial
+from documentdb_tests.framework.executor import execute_command
+from documentdb_tests.framework.parametrize import pytest_params
+
+CHANGED_DOC_TESTS: list[UpdateTestCase] = [
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_array_positional.py]: 🟠 Major — test_max_array_positional.py overlap with array operator folder

  Per FOLDER_STRUCTURE.md "Container features … one test per sub-feature, no edge cases. Edge cases belong in
  each sub-feature's own directory" and Rule 4 "Higher-level features … contain lower-level features.
  Higher-level tests only verify lower-level features are supported (1-2 cases)."

  $, $[], $[<identifier>] are array positional operators — they exist in
  tests/core/operator/update/array/positional/, positional-all/, positional-filtered/ (verified above). Per the
  rule, $max here should have 1-2 cases showing it accepts these operators, not 4 test functions covering each
  variant exhaustively (test_max_positional_changed_doc, _update_result, _filtered_positional,
  _filtered_positional_nested).

  [Suggestion] Keep one representative case per positional operator (3 cases total: $, $[], $[elem]). The
  exhaustive coverage of these positional operators belongs in update/array/positional*/, not under $max.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        {"update": collection.name, "updates": [{"q": {"_id": 1}, "u": {"$max": {"a": 50}}}]},
+    )
+    # Verify via compound index query
+    result = execute_command(collection, {"find": collection.name, "filter": {"a": 50, "b": 20}})
+    assertSuccess(
+        result, [{"_id": 1, "a": 50, "b": 20}], "Compound index should reflect updated field value"
+    )
+
+
+def test_max_large_document_near_bson_limit(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py]: 🟠 Major — test_max_edge_cases.py::test_max_large_document_near_bson_limit doesn't test what it says

  The test docstring claims "approaching 16MB BSON limit" — but the document has ~1MB string. That's not "near"
  the BSON limit; it's 6% of it. Either:
  - Make it actually near the limit (~15MB), which has real signal about $max behavior on large docs.
  - Drop it — 1MB documents are routine and don't exercise any limit.

  Also, this is performance/scale territory, not $max behavior. Probably belongs in a separate performance test
  suite, not in compatibility tests. [Suggestion] Drop it.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+Tests $max with indexed fields, deeply nested paths,
+and document validation interaction.
+"""
+
+from documentdb_tests.framework.assertions import assertSuccess
+from documentdb_tests.framework.executor import execute_command
+
+
+def test_max_on_indexed_field(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py]: Not testing $max.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    )
+    result = execute_command(collection, {"find": collection.name, "filter": {"_id": 1}})
+    assertSuccess(
+        result,
+        [{"_id": 1, "a": {"b": {"c": {"d": {"e": {"f": 42}}}}}}],
+        "Should create deeply nested path",
+    )
+
+
+def test_max_on_compound_index_field(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py]: Not testing $max.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    execute_command(
+        collection,
+        {"update": collection.name, "updates": [{"q": {"_id": 1}, "u": {"$max": {"val": 50}}}]},
+    )
+    # Verify via index scan
+    result = execute_command(collection, {"find": collection.name, "filter": {"val": 50}})
+    assertSuccess(result, [{"_id": 1, "val": 50}], "Index should reflect updated value")
+
+
+def test_max_deeply_nested_path(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py]: We just need the simple nested case. As compatibility test, we don't need test engine implementation has no bug. It works on deeply nested path is implied if simple case is supported.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_field_paths.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+import pytest
+
+from documentdb_tests.compatibility.tests.core.operator.update.utils import UpdateTestCase
+from documentdb_tests.framework.assertions import assertSuccess
+from documentdb_tests.framework.executor import execute_command
+from documentdb_tests.framework.parametrize import pytest_params
+
+_DATE = datetime(2023, 6, 15, tzinfo=timezone.utc)
+
+TESTS: list[UpdateTestCase] = [
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_field_paths.py:21]: These use $max: {"arr.0": 20} and $max: {"arr.0.field": 20}. Numeric dotted-path indexing into arrays is a
  field-path resolution rule, not a $max-specific behavior. It's the same code path that resolves $set:
  {"arr.0": 20} and any other update operator — if the path resolution works for $set, it works for $max. One
  representative case is enough; testing every path-shape per operator multiplies cases without signal.

  [Suggestion] Reduce to one path-shape case for $max. We will have path resolution as separate feature.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        collection.insert_many(test.setup_docs)
+
+    result = execute_command(
+        collection,
+        {"update": collection.name, "updates": [{"q": test.query, "u": test.update}]},
+    )
+    assertFailureCode(result, test.error_code, msg=test.msg)  # type: ignore[arg-type]
+
+
+def test_max_unique_index_conflict(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py]: Not testing $max.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    UpdateTestCase(
+        "conflict_with_inc_same_field",
+        setup_docs=[{"_id": 1, "val": 10}],
+        query={"_id": 1},
+        update={"$max": {"val": 20}, "$inc": {"val": 5}},
+        error_code=CONFLICTING_UPDATE_OPERATORS_ERROR,
+        msg="$max and $inc on same field should produce conflict error",
+    ),
+    UpdateTestCase(
+        "on_id_field_error",
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py]: 🟡 Minor — test_max_error_cases.py::on_id_field_error

  Per the comment-row: $max can't modify _id because _id is immutable. That's an immutable-field property, not a
  $max-specific behavior — every update operator hits this. Unless $max has documented special handling here,
  this is generic update-on-immutable-field coverage. [Suggestion] Verify this isn't already covered in a shared
  update-immutable-field test; if so, drop.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/test_fields_combined_operators.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+"""
+
+import pytest
+
+from documentdb_tests.compatibility.tests.core.operator.update.utils import UpdateTestCase
+from documentdb_tests.framework.assertions import assertSuccess
+from documentdb_tests.framework.executor import execute_command
+from documentdb_tests.framework.parametrize import pytest_params
+
+CHANGED_DOC_TESTS: list[UpdateTestCase] = [
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/operator/update/fields/test_fields_combined_operators.py]: We don't need all combination of these operators. We only add combined tests if there's special behavior when used with other features. $max and $set should just work in parallel, it's implied spec of update itself.

**Review by eerxuan — CHANGES_REQUESTED** [review]

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_bson_wiring.py (hunk 1 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_bson_wiring.py:18]: Fixed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py (hunk 2 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py]: Fixed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py (hunk 3 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_comparisons.py:24]: Fixed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_update_methods.py (hunk 4 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_update_methods.py]: Fixed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_array_positional.py (hunk 5 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_array_positional.py]: Fixed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py (hunk 9 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_edge_cases.py]: Removed deep nesting.

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_field_paths.py (hunk 10 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_field_paths.py:21]: Fixed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py (hunk 11 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py]: Removed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py (hunk 12 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/max/test_max_error_cases.py]: Removed

File: documentdb_tests/compatibility/tests/core/operator/update/fields/test_fields_combined_operators.py (hunk 13 above)
Comment (imforster) [reply documentdb_tests/compatibility/tests/core/operator/update/fields/test_fields_combined_operators.py]: Removed

**Review by eerxuan — COMMENTED** [review]
Three follow-up items in test_max_edge_cases.py:
  - Drop test_max_on_indexed_field (not $max-specific)
  - Drop test_max_on_compound_index_field (not $max-specific)
  - Drop or rewrite test_max_large_document_near_bson_limit (1MB ≠ "near 16MB")

  If all three are dropped, test_max_edge_cases.py becomes empty — the file itself can be removed. That's a
  valid outcome: $max edge cases (boundary numerics, special values) already live in
  test_max_numeric_comparisons.py.

**Review by eerxuan — APPROVED** [review]
